### ISSUE TITLE
jstachio logo

### ISSUE THREAD

**Issue by sviperll** [jstachio/jstachio#9]
I went ahead and drew what I imagine may be a logo for jstachio. I thought that current logo is more like a place holder. What do you think?

![jstachio-icon](https://user-images.githubusercontent.com/53340/203039101-6490e664-8bcd-4f22-84dd-dc07df13cc27.png)

**Comment by agentgt** [comment]
Someone is clearly better at drawing than me :)

Originally the logo was a placeholder particularly because I don't know how to use macSVG and could not get Inkscape to work on my m1 mac.

My critique is that I wanted to avoid the whole coffee cup thing. Most projects do these days even though it is extremely doubtful that Oracle would bring a take down order it still is a possibility.

The other reason is there is a non trivial chance I might reuse the domain name and org name to write a Typescript version that uses JSON schema as the type system. Thus a Pistachio (yes I know it doesn't look like one exactly as I suck at drawing particularly with macSVG) seem more language neutral.

And finally the Pistachio idea gave me a color scheme which I find more important than logos. The color scheme I think in web colors is "LimeGreen" and I think "Burlywood" (I might be wrong on that one). You mustaches appear to be "Peru" I think. That color might be better however I have taking a liking to LimeGreen.

I actually had started using that color scheme to CSS the Javadoc as I basically have made the generated Javadoc the official jstach.io website for now (on a side not I have even written some scripts in Java and jquery to make the javadoc more pleasant). I just don't have time to manage two documentations so I'm pushing javadoc to the limit.

All that being said I really like what you did with the "mustache" with the grooves. Can we maybe change the coffee cup out for the pistachio and give the logo a green mustache?

Also I recommend we use SVG if possible.

Thank you for the help!

**Comment by sviperll** [comment]
Heh, how about this.

![jstachio-icon2](https://user-images.githubusercontent.com/53340/203082019-f25ebea5-9bca-4202-b03c-ec3ba87bed71.png)

I will upload SVG later, I haven't done it because SVG-preview seems to wrongly render some image elements. If you agree on the image I can also adjust the colors, for now I've just used what seems to please my eye. We can also discuss other ideas for what should be in the icon. But I think I'm not available for lots of iterations, because it requires quite big amout of time...

**Comment by agentgt** [comment]
Yeah I wouldn't worry about it for now. Honestly we are already miles ahead of most projects in this space.

Like seriously look at https://github.com/spullara/mustache.java . No logo. Hardly any javadoc. Not a single coding example etc.  My wife also has some experience with logo design I just haven't asked her yet (you are better than me but I think she might be better than both of us).

What really need cycles on is integration with existing frameworks particularly Spring and if we are looking for a poster child to kind of mimic or vastly improve on is [Qute](https://quarkus.io/guides/qute) and [Jte](https://jte.gg/). Particularly Qute albeit it is mainly for Quarkus. I would like to be the goto for Spring Native instead of Thymeleaf (that is the Qute of Spring).

**Comment by agentgt** [comment]
The other major todo is better template development experience.

Right now you have two choices when you change the template:

 * Recompile and often times you need a full recompile as Eclipse and or other incremental compilers are not smart enough to know that the template is associated with some Java file.
 * Use the JMustache extension: https://jstach.io/jstachio/io.jstach.opt.jmustache/io/jstach/opt/jmustache/JMustacheRenderer.html

The JMustache extension mimics JStachio like ~ 90% (ballparking) with the giant exception that it cannot do template inheritance and I believe partial spacing might off. The JMustache extension is basically a stop gap for now. 

The two solutions I see are:

1. Dynamically programatically call some compiler while the app is running. This is far trickier than JSP, Rocker, Jte etc as we need a real Java compiler and not something that just writes byte code like ASM... and we need it to kick off the annotation processor. 
2. Make a reflection based JStachio (to replace JMustache) that is semantically runtime equivalent to compile time JStachio.

The first one might be easiest if it is possibly but it might not be. One project to look at that kind of does this is [Cute](https://github.com/toolisticon/cute) (not to be confused with qute) albeit it is doing it for unit testing. Aka programmatic kick of the annotation processing framework.

The second one although seems easier actually requires rewriting the compiler.

But the above should probably be in a separate issue altogether.

**Comment by sviperll** [comment]
This is a complex topic to explore... My quick comment is something like the following.

I would go try to do "reflection based JStachio", I don't think this is that hard to pull-off... The compiler already collects lots of metadata, that can be enough to interpret and assemble the result in the runtime.

As for programmatic compiler invocation, I don't think we need it in the recent future. But I think we can do it by using ASM and generating bytecode directly. The code generated by JStachio is quite simple, so it is possible to do direct bytecode emission...

**Comment by sviperll** [comment]
Here is the SVG to finish the thread about the logo

![jstachio-icon2](https://user-images.githubusercontent.com/53340/203289156-ed58576c-bd66-4e2d-8e2a-f6088c731bb2.svg)

**Comment by agentgt** [comment]
Thank you! I'll try to incorporate it soon. I might be offline for a week or so.

**Comment by agentgt** [comment]
Sadly I ran out of time trying to incorporate this design. While the original design is simple and looks possibly too simple it scales down nicely. I'm not sure the logo hurts adoption or really helps compared to say Spring integration or a intellij pluign.

Also this design might have too many details to work across the board but I appreciate the work put into it @sviperll .

Let me close for now and we can revisit it for 2.0.0.
